File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/roadrunner/drive/KalmanTwoWheelLocalizer.java
package org.firstinspires.ftc.teamcode.roadrunner.drive;

import androidx.annotation.NonNull;

import com.qualcomm.robotcore.hardware.HardwareMap;

import ftc.rogue.blacksmith.util.kalman.KalmanFilter;

import java.util.Arrays;
import java.util.List;

/**
 * Improve localization through running each element of localization data through Kalman Filters
 * Author: TLindauer
 */
public class KalmanTwoWheelLocalizer extends TwoWheelTrackingLocalizer {
    private final KalmanFilter
        headingFilter,
        wheelPos1Filter,
        wheelPos2Filter,
        headingVelocityFilter,
        wheelPos1VelocityFilter,
        wheelPos2VelocityFilter;

    public KalmanTwoWheelLocalizer(HardwareMap hardwareMap, SampleMecanumDrive drive) {
        super(hardwareMap, drive);

        headingFilter = new KalmanFilter(0.25, 0.125); // Remember heading is in radians
        headingVelocityFilter = new KalmanFilter(0.500, 0.225);
        wheelPos1Filter = new KalmanFilter(9, 11);
        wheelPos2Filter = new KalmanFilter(9, 11);

        wheelPos1VelocityFilter = new KalmanFilter(8, 7);
        wheelPos2VelocityFilter = new KalmanFilter(8, 7);
    }

    @Override
    public double getHeading() {
        return headingFilter.filter(super.getHeading());
    }

    @Override
    public Double getHeadingVelocity() {
        Double in = super.getHeadingVelocity();
        if (in != null)
            return headingVelocityFilter.filter(in);
        return 0.0;
    }

    @NonNull
    @Override
    public List<Double> getWheelPositions() {
        List<Double> in = super.getWheelPositions();
        return Arrays.asList(wheelPos1Filter.filter(in.get(0)), wheelPos2Filter.filter(in.get(1)));
    }

    @NonNull
    @Override
    public List<Double> getWheelVelocities() {
        List<Double> in = super.getWheelVelocities();
        return Arrays.asList(wheelPos1VelocityFilter.filter(in.get(0)), wheelPos2VelocityFilter.filter(in.get(1)));
    }
}

File: Blacksmith/src/main/kotlin/ftc/rogue/blacksmith/util/kalman/KalmanTwoWheelLocalizer.kt
package ftc.rogue.blacksmith.util.kalman

import com.acmerobotics.roadrunner.geometry.Pose2d
import com.acmerobotics.roadrunner.localization.TwoTrackingWheelLocalizer
import ftc.rogue.blacksmith.units.AngleUnit.DEGREES
import ftc.rogue.blacksmith.util.toRad

class KalmanTwoWheelLocalizer(
    private val localizer: TwoTrackingWheelLocalizer,
) : TwoTrackingWheelLocalizer(localizer.getWheelPoses()) {

    // Remember heading is in radians
    private val headingFilter = KalmanFilter(0.25, 0.125)
    private val wheelPos1Filter = KalmanFilter(9.0, 11.0)
    private val wheelPos2Filter = KalmanFilter(9.0, 11.0)
    private val headingVelocityFilter = KalmanFilter(0.500, 0.225)
    private val wheelPos1VelocityFilter = KalmanFilter(8.0, 7.0)
    private val wheelPos2VelocityFilter = KalmanFilter(8.0, 7.0)

    override fun getHeading(): Double {
        return headingFilter.filter(localizer.getHeading())
    }

    override fun getWheelPositions(): List<Double> {
        val poses = localizer.getWheelPositions()

        return listOf(
            wheelPos1Filter.filter(poses[0]),
            wheelPos2Filter.filter(poses[1]),
        )
    }

    override fun getWheelVelocities(): List<Double> {
        val poses = localizer.getWheelVelocities()!!

        return listOf(
            wheelPos1VelocityFilter.filter(poses[0]),
            wheelPos2VelocityFilter.filter(poses[1]),
        )
    }

    override fun getHeadingVelocity(): Double {
        val velocity = localizer.getHeadingVelocity()

        return velocity
            ?.let { headingVelocityFilter.filter(it) }
            ?: 0.0
    }

    fun setHeadingFilterCoeffs(R: Double, Q: Double) = apply {
        headingFilter.setProcessNoise(R)
        headingFilter.setMeasurementNoise(Q)
    }

    fun setWheelPos1FilterCoeffs(R: Double, Q: Double) = apply {
        wheelPos1Filter.setProcessNoise(R)
        wheelPos1Filter.setMeasurementNoise(Q)
    }

    fun setWheelPos2FilterCoeffs(R: Double, Q: Double) = apply {
        wheelPos2Filter.setProcessNoise(R)
        wheelPos2Filter.setMeasurementNoise(Q)
    }

    fun setHeadingVelocityFilterCoeffs(R: Double, Q: Double) = apply {
        headingVelocityFilter.setProcessNoise(R)
        headingVelocityFilter.setMeasurementNoise(Q)
    }

    fun setWheelPos1VelocityFilterCoeffs(R: Double, Q: Double) = apply {
        wheelPos1VelocityFilter.setProcessNoise(R)
        wheelPos1VelocityFilter.setMeasurementNoise(Q)
    }

    fun setWheelPos2VelocityFilterCoeffs(R: Double, Q: Double) = apply {
        wheelPos2VelocityFilter.setProcessNoise(R)
        wheelPos2VelocityFilter.setMeasurementNoise(Q)
    }

    // -- INTERNAL --

    companion object {
        private fun TwoTrackingWheelLocalizer.getWheelPoses(): List<Pose2d> {
            val parallelX = this::class.java.getField("PARALLEL_X").get(null) as Double
            val parallelY = this::class.java.getField("PARALLEL_Y").get(null) as Double
            val perpendicularX = this::class.java.getField("PERPENDICULAR_X").get(null) as Double
            val perpendicularY = this::class.java.getField("PERPENDICULAR_Y").get(null) as Double

            return listOf(
                Pose2d(parallelX, parallelY, 0.0),
                Pose2d(perpendicularX, perpendicularY, 90.toRad(from = DEGREES)),
            )
        }
    }
}
